src/01-function-calling/index.ts

```diff
@@ -0,0 +1,248 @@
+ *
+ * Usage examples:
+ *
+ * bun run src/01-function-calling/index.ts "who is popular on artsy right now?" | jq
+ * bun run src/01-function-calling/index.ts "who is artsy into these days?" | jq
+ * bun run src/01-function-calling/index.ts "who is new on artsy?" | jq
+ * bun run src/01-function-calling/index.ts "give me the last 17 artists on artsy, alphabetically" | jq
+ */
+
+import dotenv from "dotenv"
+import OpenAI from "openai"
+
+dotenv.config()
+const openai = new OpenAI() // uses `OPENAI_API_KEY` from .env
+
+/*
+ * Get user input
+ */
+
+const input =
+  process.argv.slice(2).join(" ") || "Who's trending on Artsy right now?"
+
+/*
+ * Create the message payload for the chat completion
+ */
+
+const messages: OpenAI.Chat.Completions.ChatCompletionMessageParam[] = [
+  {
+    role: "system",
+    content:
+      `You are a helpful assistant that can provide information about the art world via Artsy's platform. ` +
+      ` ` +
+      `If you receive a question that cannot be answered with one of your known tools, do not force a tool response. ` +
+      `Instead ask for clarification or say that you don't know how to answer the question. ` +
+      ` `,
+  },
+  {
+    role: "user",
+    content: input,
+  },
+]
+
+/*
+ * Define the tools that the chat completion can use:
+ *
+ * 1. get_artists: Get a list of artists on Artsy
+ * 2. get_curated_artists: Get a list of curated artists on Artsy
+ */
+
+const tools: OpenAI.Chat.Completions.ChatCompletionTool[] = [
+  {
+    type: "function",
+    function: {
+      name: "get_artists",
+      description: `Get a list of artists on Artsy. Artists may be sorted chronologically by creation date, alphabetically by name, or in descending order of a popularity/trending score.`,
+      parameters: {
+        type: "object",
+        properties: {
+          size: {
+            type: "integer",
+            description: "The number of artists to return",
+            default: 5,
+            minimum: 1,
+            maximum: 20,
+          },
+          sort: {
+            type: "string",
+            description: "The sort order in which to return artists",
+            default: "SORTABLE_ID_ASC",
+            enum: [
+              "CREATED_AT_ASC",
+              "CREATED_AT_DESC",
+              "SORTABLE_ID_ASC",
+              "SORTABLE_ID_DESC",
+              "TRENDING_DESC",
+            ],
+          },
+        },
+      },
+    },
+  },
+  {
+    type: "function",
+    function: {
+      name: "get_curated_artists",
+      description: `Get a list of curated artists on Artsy. These are artists whose works have been highlighted by Artsy curators, and may change from week to week.`,
+      parameters: {
+        type: "object",
+        properties: {
+          size: {
+            type: "integer",
+            description: "The number of artists to return",
+            default: 5,
+            minimum: 1,
+            maximum: 20,
```

Comment:
Does the model enforce these constraints, or could I override them via a user prompt? Something like "give me 100 curated artists?"

Reply:
Speaks to the importance of validating the function call signature before actually doing anything with it.

Reply:
💯